### ISSUE TITLE
Hide Mouse Cursor

### ISSUE THREAD

**Issue by shadow111** [robmikh/Win32CaptureSample#28]
Hello,
is it possible to wrap this code to be called from JavaScript, we are interested to export the hide cursor from capture function but it seems that we can't exported it alone it should be attached to a capture session.
I just wanna to know if it's possible to export the hide cursor function alone.
Best Regards
Lafi RAED

**Comment by robmikh** [comment]
I'm a little lost. What do you mean about exporting the hide cursor function alone? It's only useful in the context of having a capture session running. What is that you're trying to do?

As for wrapping this to be called in JavaScript, I don't have much experience with that. It will be up to the JavaScript engine that you use. I would start by putting all the capture code into a dll with a simplified C-style ABI. From there, there should be plenty of resources online on how to call a DLL from your JavaScript engine of choice.

**Comment by shadow111** [comment]
Hello and thank you for your reply, yes i start doing like that.
How i can just stream  the frame buffer  from the App.cpp, it's seem that the frame are   inside a  lambda

**Comment by robmikh** [comment]
There won't be any bits because you aren't waiting for the frame to come in. When your function returns the session is torn down and the capture is ended.

Pass an event handle along to your lambda and signal it when you're done (instead of the `completion.set(result)` line). After you call `StartCapture`, wait on the event. You should have your bits after that.

You also may end up copying data that isn't part of the image. Often times the image will be contained in more video memory than the reported size of the buffer. Take a look at [this comment](https://github.com/robmikh/Win32CaptureSample/issues/25#issuecomment-718265205) from another issue for some code that does the copy differently.

**Comment by shadow111** [comment]
thank you for your reply
what i want to do is to return every frame bytes  while the capture session is not stopped.
can you please show me  an example with event handler so the capture will not be stoped after the return.
Im not asking for a complete code just to put me on the right direction

**Comment by robmikh** [comment]
If you only want to capture a single frame, take a look at this file:

https://github.com/robmikh/ScreenshotFun/blob/c022b523b9e849f782e5046891500998645e098b/ScreenshotFun/CaptureSnapshot.cpp#L25

The key point is that it waits on an event handle, to make sure you get a frame back before returning:

```cpp
session.StartCapture();

// Don't return until the capture is finished
captureEvent.wait();
WINRT_ASSERT(result != nullptr);

return result;
```

If you want to capture multiple frames, you will want to keep a reference to both the `GraphicsCaptureSession` as well as the `Direct3D11CaptureFramePool` like the `SimpleCapture` object does in this sample:

https://github.com/robmikh/Win32CaptureSample/blob/master/Win32CaptureSample/SimpleCapture.cpp
https://github.com/robmikh/Win32CaptureSample/blob/master/Win32CaptureSample/SimpleCapture.h

When you're done, call `Close` on both.

**Comment by shadow111** [comment]
Thank you for your reply, really appreciate that
on the case of multiple frame which will allow us to make screen recording the reference should be kept here  `framePool.FrameArrived` based on SimpleCapture example i really cannot get it very clear since in the code you are referencing it on the class constructor `m_framePool.FrameArrived({ this, &SimpleCapture::OnFrameArrived });` and in the case of single frame you already kept the reference as well `framePool.FrameArrived([session, d3dDevice, d3dContext, &result, captureEvent](auto& framePool, auto&)`  is possible to provide me with a sample that does the same thing , i have been struggling and a bit confused on it and i can't get it right . 
I really appreciate that

**Comment by robmikh** [comment]
No, the CaptureGifEncoder project absolutely holds a reference to both the `GraphicsCaptureSession` and the `Direct3D11CaptureFramePool`. As long as `std::getline` doesn't return, the reference is kept. Once it returns, the code then releases the reference and by then the gif is done encoding. 

I would strongly recommend you familiarize yourself with [COM](https://docs.microsoft.com/en-us/windows/win32/com/component-object-model--com--portal) and how `AddRef` and `Release` work before embarking on a project like this. Particularly you should take a look at [this article from MSDN](https://docs.microsoft.com/en-us/windows/win32/com/managing-object-lifetimes-through-reference-counting). You'll notice that `AddRef` and `Release` are not called explicitly, but rather are manged through the smart pointer provided by C++\WinRT (e.g. `winrt::com_ptr<T>`, or the projected runtime classes like `GraphicsCaptureSession` and `Direct3D11CaptureFramePool`). 

But at this point I think this issue is outside the bounds of this sample. I'm going to close this issue, but I'll leave you with topics/steps to investigate:

1. Take the core of `SimpleCapture` and expose it as some opaque pointer in a C ABI.
2. Export a C function that takes the pointer from step 1 and a function pointer, which will be internally stored.
3. In `OnFrameArrived`, extract the bytes from the frame and then call the function pointer with the array of bytes and some basic information.
4. Export a C function that takes the pointer from step 1, which will internally close the object and end the capture.

At that point, it'll be up to you to glue this together with a native nodeJS plug-in. Best of luck!
